### ISSUE TITLE
WEBGL Context Lost

### ISSUE THREAD

**Issue by emrgry** [godotengine/godot#65696]
### Godot version

3.5 stable

### System information

macOS Monterey version 12.5.1 , MacBook Pro(13-inch,M1,2020)

### Issue description

My HTML5 export works on windows but doesn't work on macOS. I tried for Opera, Chrome and Mozilla browsers and it gave me this error . 

<img width="447" alt="Screen Shot 2022-09-12 at 15 16 35" src="https://user-images.githubusercontent.com/32719145/189652143-116cfa66-f981-4458-ab9c-4416c4c39526.png">

When I checked the console, I saw this:
<img width="1371" alt="Screen Shot 2022-09-12 at 15 32 50" src="https://user-images.githubusercontent.com/32719145/189654492-dc5629ae-74f9-4579-a7fb-94134d228ac6.png">

You can check my log file
[127.0.0.1-1662985419802.log](https://github.com/godotengine/godot/files/9547956/127.0.0.1-1662985419802.log)

### Steps to reproduce

<img width="1371" alt="Screen Shot 2022-09-12 at 15 38 32" src="https://user-images.githubusercontent.com/32719145/189655589-a2ddfec3-21ec-473f-8b2d-a8da34e1347f.png">

console shows this URL: [https://bugs.freedesktop.org/show_bug.cgi?id=100316](url) You can check here for reproduce #

**Comment by Calinou** [comment]
Try using the GLES2 rendering backend instead of GLES3. There are lots of known issues with 3.x's GLES3 backend in HTML5.

**Comment by emrgry** [comment]
I have to use spotlight in my project. As I know GLES 2 doesn't support spotlight

**Comment by akien-mga** [comment]
We need to fix the WebGL shader error reporting. In the logs it properly included the full shader content that failed to compile, but it did not include the line on which compilation fails.

**Comment by emrgry** [comment]
> > I have to use spotlight in my project. As I know GLES 2 doesn't support spotlight
> 
> GLES2 does support SpotLights, even with shadows enabled. If SpotLights doesn't work in HTML5 only, then that's a bug that should be reported separately.
> 
> There is an issue reporting shadows not working in HTML5, but that's in GLES3: #57500

GLES 2 works for now thanks.

**Comment by OwlyCode** [comment]
I had the same kind of issue with my game made with godot 4. I noticed that once the alert for context loss is closed, I could hear the music of the game and even play it (blindly, with no image). Once I reloaded the page and started the game again, it worked.

Exiting chrome and restarting it re-created the bug.

(Here is the game: https://owlycode.itch.io/escape-the-breach)

**Comment by lpares12** [comment]
Having similar issues in Godot 4:

Having issues too. My project is set to "Mobile" and there are so many errors related to shdaer_gles3.cpp on the developer's console in Chrome. Same project works fine on other OSs.

This is the output in the console:

<img width="555" alt="217590849-5823e014-2f81-4cf8-9012-ef61b5af0419" src="https://user-images.githubusercontent.com/9217676/218973149-0770cfd9-0325-4699-af42-4a1c97c51b21.png">
<img width="555" alt="217590938-7da9e732-e646-4916-a0ff-0d0313e59368" src="https://user-images.githubusercontent.com/9217676/218973169-a791e0cd-c9bb-45e3-8fdc-c5ad2ad0fbc3.png">
<img width="555" alt="217590955-54f9d688-8d81-4a9b-a734-92fe820900c8" src="https://user-images.githubusercontent.com/9217676/218973181-b7e05826-3ca6-4395-be90-6654b3f017b3.png">

Also note that in one of the 4 macOS laptops I tried, it did work but took 3-4 minutes to load.

**Comment by Calinou** [comment]
> My project is set to "Mobile" and there are so many errors related to shdaer_gles3.cpp on the developer's console in Chrome. Same project works fine on other OSs.

Forward Mobile is only usable on desktop and mobile platforms, not HTML5. On HTML5, `gl_compatibility` is always used.

**Comment by lpares12** [comment]
@Calinou I see, so I guess changing it to Compatibility will have exactly the same outcome. Or should I give it a try anyways?

**Comment by Calinou** [comment]
> @Calinou I see, so I guess changing it to Compatibility will have exactly the same outcome. Or should I give it a try anyways?

It will behave exactly the same on HTML5, so there's no point.

**Comment by moribellamy** [comment]
> We need to fix the WebGL shader error reporting. In the logs it properly included the full shader content that failed to compile, but it did not include the line on which compilation fails.

Perhaps we are still blocked on needing better instrumentation? Is there a bug open for that?

Throwing in my repro in case it helps down the line:

source: https://github.com/moribellamy/platonic-game/commit/fb72372b585cb56b61ef509b06bec555028b366e
deployed: https://razielx.itch.io/godot-shader-crash

Does the Godot steering committee plan to support Web as a first class compilation target for 4.x? I ask because it seems like there might be some deep issues here?

**Comment by Calinou** [comment]
> Perhaps we are still blocked on needing better instrumentation? Is there a bug open for that?

There's no specific issue open for it; I suggest opening a separate issue to track this.

> Does the Godot steering committee plan to support Web as a first class compilation target for 4.x?

There's no steering committee to speak of. :slightly_smiling_face: 

In open source, people work on what they want to work on, so there's no ETA for resolving this issue. We intend to keep supporting the web export, but we have relatively few contributors available to work on it. The web platform is also prone to a lot of third-party issues that contributors generally struggle dealing with, because they require specific hardware/software configurations (namely, macOS/iOS-only issues).

**Comment by moribellamy** [comment]
> There's no steering committee to speak of.

> We intend to keep supporting the web export

What is the antecedent of 'We'? You understand my confusion 😵‍💫 .

I'm glad to help to the limited extend I am able, but I'm just trying not to do a ton of work that gets rejected for political reasons.

**Comment by Calinou** [comment]
I can't reproduce this issue in an empty project on a M1 Mac mini (macOS 13, Chromium 118), but we need a project that is as minimal as possible but still reproduces this issue. The project linked in https://github.com/godotengine/godot/issues/65696#issuecomment-1793055484 isn't minimal enough to really point out at what is exactly causing this issue.

**Comment by moribellamy** [comment]
It's definitely the GPUParticles2D. This patch makes the error go away:
```diff --git a/enemies/slime/slime.gd b/enemies/slime/slime.gd
index 4714111..57c3525 100644
--- a/enemies/slime/slime.gd
+++ b/enemies/slime/slime.gd
@@ -71,7 +71,7 @@ func _on_hurt_box_hurt(hitbox: HitBox) -> void:
        process_material.direction.x = -abs(process_material.direction.x) \
                if source_is_to_right \
                else abs(process_material.direction.x)
-       LifecycleHelpers.show(_hurt_particles, _hurt_particles.lifetime)
+       #LifecycleHelpers.show(_hurt_particles, _hurt_particles.lifetime)

        if _hp <= 0:
                LifecycleHelpers.play_audio(_die_sound)```

I haven't done anything too tricky. Maybe GPUParticles2D don't work at all on the web?
![Screenshot 2023-11-05 at 3 35 26 PM](https://github.com/godotengine/godot/assets/13877974/f6edcec7-a894-47db-ab72-608c8f1acd3a)

Not sure if I'll have time to give a smaller project anytime soon.

**Comment by Calinou** [comment]
A similar issue was fixed by https://github.com/godotengine/godot/issues/75270, but the error message you're getting is different this time.

cc @cody82

**Comment by moribellamy** [comment]
IDK how shader code works, but it seems that `_compile_specialization()` calls `_build_variant_code()` in on demand, at runtime. in my case, it appears the following shader was built, and it crashed the browser.

[badshader.glsl.txt](https://github.com/godotengine/godot/files/13324566/badshader.glsl.txt)

**Comment by moribellamy** [comment]
the shader godot tried to compile at runtime (linked above) declared `#version 300`, but it also tries to execute `layout(location = N) ...`, which is a feature added in 330. so that cannot be helping.

i would think any shader compiled this way would lead to undefined behavior on any platform. @Calinou do you know anyone who can confirm or deny?
